src/viur/core/modules/user.py

```diff
@@ -616,133 +619,211 @@ def login(self, skey="", token="", *args, **kwargs):
+        # Check if the OTP device has a time drift
+
+        timedriftchange = float(res) - otp_user_conf["timedrift"]
+        if abs(timedriftchange) > 2:
+            # The time-drift change accumulates to more than 2 minutes (for interval==60):
+            # update clock-drift value accordingly
+            self.updateTimeDrift(user_key, timedriftchange)
+
+        # Continue with authentication
+        return self.userModule.secondFactorSucceeded(self, user_key)
+
+    @staticmethod
+    def verify(
+        otp: str | int,
+        secret: str,
+        algorithm: str = "sha1",
+        interval: int = 60,
+        timedrift: float = 0.0,
+        for_time: datetime.datetime | None = None,
+        valid_window: int = 0,
+    ) -> int | None:
+        """
+        Verifies the OTP passed in against the current time OTP.
+
+        This is a fork of pyotp.verify. Rather than true/false, if valid_window > 0, it returns the index for which
+        the OTP value obtained by pyotp.at(for_time=time.time(), counter_offset=index) equals the current value shown
+        on the hardware token generator. This can be used to store the time drift of a given token generator.
+
+        :param otp: the OTP token to check against
+        :param secret: The OTP secret
+        :param algorithm: digest function to use in the HMAC (expected to be sha1 or sha256)
+        :param interval: the time interval in seconds for OTP. This defaults to 60 (old OTP c200 Generators). In
+        pyotp, default is 30!
+        :param timedrift: The known timedrift (old index) of the hardware OTP generator
+        :param for_time: Time to check OTP at (defaults to now)
+        :param valid_window: extends the validity to this many counter ticks before and after the current one
+        :returns: The index where verification succeeded, None otherwise
+        """
+        # get the hashing digest
+        digest = {
+            "sha1": hashlib.sha1,
+            "sha256": hashlib.sha256,
+        }.get(algorithm)
+
+        if not digest:
+            raise errors.NotImplemented(f"{algorithm=} is not implemented")
+
+        if for_time is None:
+            for_time = datetime.datetime.now()
+
+        # Timedrift is updated only in fractions in order to prevent problems, but we need an integer index
+        timedrift = round(timedrift)
+        secret = bytes.decode(base64.b32encode(bytes.fromhex(secret)))  # decode secret
+        otp = str(otp).zfill(6)  # fill with zeros in front
+
+        # logging.debug(f"TimeBasedOTP:verify: {digest=}, {interval=}, {valid_window=}")
+        totp = pyotp.TOTP(secret, digest=digest, interval=interval)
+
+        if valid_window:
+            for i in range(timedrift - valid_window, timedrift + valid_window + 1):
+                token = str(totp.at(for_time, i))
+                # logging.debug(f"TimeBasedOTP:verify: {i=}, {otp=}, {token=}")
+                if hmac.compare_digest(otp, token):
+                    return i
```

Comment:
Can you rename the `i` to a more meaningful variable?

I suggest `offset`, because it's the short form of the second parameter of [`pyotp.totp.TOTP.at()`](https://pyauth.github.io/pyotp/#pyotp.totp.TOTP.at)

```suggestion
            for offset in range(timedrift - valid_window, timedrift + valid_window + 1):
                token = str(totp.at(for_time, offset))
                # logging.debug(f"TimeBasedOTP:verify: {offset=}, {otp=}, {token=}")
                if hmac.compare_digest(otp, token):
                    return offset
```